### PR TITLE
SQL: Fix SQL dataframe resampling (fill mode + time intervals)

### LINKED ISSUE #36627: $__timeGroup changed behaviour in 8.x
`$__timeGroup(_, _, 0)` changed its behaviour from `7.5.9` to `8.0.5` and I do not know how to get the old behaviour back other than downgrading.

I have a graph with PostgreSQL as data source. A sample query:

```
 WITH data AS (SELECT now()-'3m'::interval AS ts, 'metric' AS metric, 42 AS n) SELECT $__timeGroup(ts, '1m', 0), metric, n FROM data
```
`7.5.9` result:
```
"Time","metric"
2021-07-11 10:02:00,0
2021-07-11 10:03:00,0
2021-07-11 10:04:00,42
2021-07-11 10:05:00,0
2021-07-11 10:06:00,0
2021-07-11 10:07:00,0
```
There is a single 42 spike, the rest is zero.

`8.0.5` result:
```
"time","metric"
2021-07-11 10:10:20,0
2021-07-11 10:11:20,42
2021-07-11 10:12:20,42
2021-07-11 10:13:20,42
2021-07-11 10:14:20,42
2021-07-11 10:15:20,42
```
The metric remains at 42 until the end of the time range.

How can I get 7.x-like results with 8.x?

### PR DESCRIPTION
**Pull request by idafurjes**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #36627 and #36878

**Special notes for your reviewer**:

### CODE REVIEW

**Review by wbrowne — CHANGES_REQUESTED** [review]
Think we're still missing some pieces here. I noticed that in the current code we only check for `fill` when `qm.Format == dataQueryFormatSeries` - see here https://github.com/grafana/grafana/blob/main/pkg/tsdb/sqleng/sql_engine.go#L280

~That doesn't seem correct to me (unless we're filling elsewhere) - so I think that block should be moved outside of there.~
Is actually correct - the way I was testing lead me to think otherwise 🤦 

At that point we're still left with different results:

7.5.x:

Screenshot 2021-07-20 at 16 17 49


Branch:

image

Also note the different column numbers, I would have expected 3 in 7.5.x too 🤔

**Comment by ying-jeanne** [comment]
tested with v8 and v7.5.0 now the behavior stay the same
Screenshot 2021-07-27 at 13 29 04
Screenshot 2021-07-27 at 13 43 55

File: pkg/tsdb/sqleng/resample.go [hunk 1]
```diff
@@ -110,17 +110,15 @@ func resample(f *data.Frame, qm dataQueryModel) (*data.Frame, error) {
 				return f, fmt.Errorf("time point is nil")
 			}
 
-			if t.(time.Time).After(currentTime) {
-				nextTime := currentTime.Add(qm.Interval)
-				if t.(time.Time).Before(nextTime) {
+			previousTime := currentTime.Add(-qm.Interval)
```
Comment (wbrowne) [inline pkg/tsdb/sqleng/resample.go:113]: mind blown

Can we add a comment to this? 😅

**Comment by wbrowne** [comment]
@ying-jeanne Still seems like something is missing:

Branch:
Screenshot 2021-07-28 at 10 54 09

(the response data stays the same no matter the time range)

7.X:
Screenshot 2021-07-28 at 11 01 51

**Review by ying-jeanne — APPROVED** [review]

**Review by wbrowne — APPROVED** [review]
Tested and it LGTM!

**Pull request merged**